File: pkg/contract/game_engine.go
// SPDX-License-Identifier: AGPL-3.0-only

package contract

import "context"

// GameEngine is the boundary for native engines and delegated upstream
// backends. Commands must be idempotent and state-aware because UI retries,
// runtime recovery, and external backend restarts are expected.
type GameEngine interface {
	Describe(context.Context) (GameEngineDescriptor, error)
	ResolveProfile(context.Context, ProfileID) (ProfileSummary, error)
	Status(context.Context, ProfileID) (RuntimeStatus, error)
	Start(context.Context, RuntimeCommand) (CommandResult, error)
	Stop(context.Context, RuntimeCommand) (CommandResult, error)
	Restart(context.Context, RuntimeCommand) (CommandResult, error)
	Refresh(context.Context, RuntimeCommand) (CommandResult, error)
	SubmitTask(context.Context, TaskRequest) (TaskRunSummary, error)
	Scheduler(context.Context, ProfileID) (SchedulerSummary, error)
	RecentLogs(context.Context, RecentQuery) ([]LogEvent, error)
	ResourceHistory(context.Context, ResourceHistoryQuery) ([]ResourceHistoryPoint, error)
	RecentAcquisitions(context.Context, AcquisitionQuery) ([]AcquisitionCapture, error)
}

type GameEngineDescriptor struct {
	ID                   string              `json:"id"`
	Kind                 EngineKind          `json:"kind"`
	Game                 GameKey             `json:"game"`
	SupportedServers     []ServerKey         `json:"supportedServers"`
	SupportedResolutions []Resolution        `json:"supportedResolutions"`
	Capabilities         []RuntimeCapability `json:"capabilities,omitempty"`
	Version              string              `json:"version,omitempty"`
}

type RuntimeCommand struct {
	ProfileID ProfileID         `json:"profileId"`
	RequestID string            `json:"requestId"`
	Reason    string            `json:"reason,omitempty"`
	Options   map[string]string `json:"options,omitempty"`
}

type CommandResult struct {
	RequestID string        `json:"requestId"`
	State     RuntimeState  `json:"state"`
	Accepted  bool          `json:"accepted"`
	Message   string        `json:"message,omitempty"`
	Error     *RuntimeError `json:"error,omitempty"`
}

type TaskRequest struct {
	ProfileID ProfileID         `json:"profileId"`
	RequestID string            `json:"requestId"`
	TaskID    TaskID            `json:"taskId"`
	FlowID    string            `json:"flowId"`
	Options   map[string]string `json:"options,omitempty"`
}

type TaskRunSummary struct {
	TaskRunID TaskRunID     `json:"taskRunId"`
	TaskID    TaskID        `json:"taskId"`
	ProfileID ProfileID     `json:"profileId"`
	State     RuntimeState  `json:"state"`
	StartedAt string        `json:"startedAt"`
	EndedAt   string        `json:"endedAt,omitempty"`
	LastError *RuntimeError `json:"lastError,omitempty"`
}

type RecentQuery struct {
	ProfileID ProfileID `json:"profileId,omitempty"`
	Limit     int       `json:"limit,omitempty"`
}

type ResourceHistoryQuery struct {
	ProfileID ProfileID   `json:"profileId,omitempty"`
	Key       ResourceKey `json:"key,omitempty"`
	Limit     int         `json:"limit,omitempty"`
}

type AcquisitionQuery struct {
	ProfileID ProfileID `json:"profileId,omitempty"`
	TaskRunID TaskRunID `json:"taskRunId,omitempty"`
	Limit     int       `json:"limit,omitempty"`
}

File: crates/actingcommand-contract/src/game_engine.rs
// SPDX-License-Identifier: AGPL-3.0-only

//! Rust mainline boundary for native engines and delegated upstream backends.

use crate::types::*;
use serde::{Deserialize, Serialize};

pub trait GameEngine {
    fn describe(&self, ctx: &RuntimeContext) -> ContractResult<GameEngineDescriptor>;

    fn resolve_profile(
        &self,
        ctx: &RuntimeContext,
        profile_id: ProfileId,
    ) -> ContractResult<ProfileSummary>;

    fn status(&self, ctx: &RuntimeContext, profile_id: ProfileId) -> ContractResult<RuntimeStatus>;

    fn start(
        &mut self,
        ctx: &RuntimeContext,
        command: RuntimeCommand,
    ) -> ContractResult<CommandResult>;

    fn stop(
        &mut self,
        ctx: &RuntimeContext,
        command: RuntimeCommand,
    ) -> ContractResult<CommandResult>;

    fn restart(
        &mut self,
        ctx: &RuntimeContext,
        command: RuntimeCommand,
    ) -> ContractResult<CommandResult>;

    fn refresh(
        &mut self,
        ctx: &RuntimeContext,
        command: RuntimeCommand,
    ) -> ContractResult<CommandResult>;

    fn submit_task(
        &mut self,
        ctx: &RuntimeContext,
        request: TaskRequest,
    ) -> ContractResult<TaskRunSummary>;

    fn scheduler(
        &self,
        ctx: &RuntimeContext,
        profile_id: ProfileId,
    ) -> ContractResult<SchedulerSummary>;

    fn recent_logs(
        &self,
        ctx: &RuntimeContext,
        query: RecentQuery,
    ) -> ContractResult<Vec<LogEvent>>;

    fn resource_history(
        &self,
        ctx: &RuntimeContext,
        query: ResourceHistoryQuery,
    ) -> ContractResult<Vec<ResourceHistoryPoint>>;

    fn recent_acquisitions(
        &self,
        ctx: &RuntimeContext,
        query: AcquisitionQuery,
    ) -> ContractResult<Vec<AcquisitionCapture>>;
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GameEngineDescriptor {
    pub id: String,
    pub kind: EngineKind,
    pub game: GameKey,
    pub supported_servers: Vec<ServerKey>,
    pub supported_resolutions: Vec<Resolution>,
    pub capabilities: Vec<RuntimeCapability>,
    pub version: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RuntimeCommand {
    pub profile_id: ProfileId,
    pub request_id: String,
    pub reason: Option<String>,
    pub options: Metadata,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CommandResult {
    pub request_id: String,
    pub state: RuntimeState,
    pub accepted: bool,
    pub message: Option<String>,
    pub error: Option<RuntimeError>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TaskRequest {
    pub profile_id: ProfileId,
    pub request_id: String,
    pub task_id: TaskId,
    pub flow_id: String,
    pub options: Metadata,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TaskRunSummary {
    pub task_run_id: TaskRunId,
    pub task_id: TaskId,
    pub profile_id: ProfileId,
    pub state: RuntimeState,
    pub started_at: Timestamp,
    pub ended_at: Option<Timestamp>,
    pub last_error: Option<RuntimeError>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RecentQuery {
    pub profile_id: Option<ProfileId>,
    pub limit: Option<i32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ResourceHistoryQuery {
    pub profile_id: Option<ProfileId>,
    pub key: Option<ResourceKey>,
    pub limit: Option<i32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AcquisitionQuery {
    pub profile_id: Option<ProfileId>,
    pub task_run_id: Option<TaskRunId>,
    pub limit: Option<i32>,
}
